### ISSUE TITLE
Port GRNsight to D3.js version 4

### ISSUE THREAD

**Issue by kdahlquist** [dondi/GRNsight#523]
Based on the conclusions from #520, we will go ahead and transition to D3.js version 4.
- Switch underlying library to v4
- Make simple search-replace changes
- Adapt zooming work to v4
- Dive into deeper force simulation changes

(depending on the size of these subtasks, we might break them up into separate issues later)

**Comment by kdahlquist** [comment]
Let's finish #248 first, then go into this.

**Comment by yshin4** [comment]
Worked on this converting issue with @eileenchoe during our work session today. I am hoping to make further progress on it on Thursday, after looking into #188.

**Comment by eileenchoe** [comment]
Today, completed first two bullets (switch underlying library & search-replace changes), and began work on updating to the new force library. All work for this issue is being done on the `d3v4` branch. Here is a beginning of a list of some tasks that still need to be finished: 
- Write force simulation for gravity, which is no longer a method of force
- Re-enable slider functionality
- Re-enable drag functionality

**Comment by kdahlquist** [comment]
Before writing our own force simulation for gravity, it is worth having a discussion as to whether we need this force at all.  We got it "for free" with D3v3, but we should assess it's usefulness before spending time there.

**Comment by eileenchoe** [comment]
Met with @dondi today about how to approach the transition to the d3v4 force library. Identified that the force object is being passed into the sliders via `sliderController.addForce()`. A first step to progress is to focus on one 'simple' force (linkDistance) and getting the interaction with the slider working.

**Comment by eileenchoe** [comment]
This is cool! I made my own [bl.ock](https://bl.ocks.org/eileenchoe/raw/767716f14eeaa586fb315f1ac6f4032d/
) with some d3v4 features I'm testing.

Figured out how to manipulate force parameters (and node dragging/freezing) for d3v4, working off of [this example](https://bl.ocks.org/mbostock/2675ff61ea5e063ede2b5d63c08020c7). At the meeting today, I want to figure out which force graph parameters we intend on keeping (or parameters we might want to add) to our sliders controller.

**Comment by dondi** [comment]
Decision from the meeting today, after reviewing the meaning of charge distance and gravity, is to go ahead and remove those sliders. We will not try to write custom versions of these forces for D3v4. Charge distance is somewhat duplicated by link distance + charge, and gravity can be superseded by the recently implemented Center button.

**Comment by kdahlquist** [comment]
Since @yshin4's issues are now closed, she should team up with @eileenchoe on the D3v4 conversion because that needs to be done before her work can start on the new graph layouts.

**Comment by eileenchoe** [comment]
During our work meeting @yshin4 and I sync'ed up about the status and scope of this issue. We will working on the `d3v4` branch. We think that it will be best to initially work together to port the core force library functions, then split out the work (along `sliders.js` and `graph.js` might be a natural separation)

**Comment by eileenchoe** [comment]
Re-enabled slider controller's ability to modify force graph parameters of "link distance" and "charge." In today's meeting will need to discuss new default values and ranges for these sliders. Still working on reset and undo reset, hope to make progress on this before the meeting.

**Comment by dondi** [comment]
Demo file issue has been fixed; the fix was already in `beta` and @eileenchoe just needed to merge `beta` into `d3v4`.

**Comment by dondi** [comment]
For slider values, we will wait until after the refactor is completed before determining the new defaults (if at all).

**Comment by yshin4** [comment]
When clicked on the graph (to drag the whole graph) there is an error regarding zoom. So I tried to change that in the graph.js file and now we do not get any errors but the graph still does not move as it should. There is no zoom.translate in v4, which we previously used in v3. Instead, there's zoom.translateBy and zoom.translateTo.

**Comment by eileenchoe** [comment]
Completed refactor of sliders object. Working with @yshin4 now on zoom. Meeting with @dondi today regarding implementing `zoom.translate` in v4.

**Comment by eileenchoe** [comment]
Finally got the interaction between graph dragging / slider zoom / d-pad control / viewport size changing / adaptive toggle (mostly) working with v4! Here are the remaining tasks @yshin4 let me know if there's something in particular you want to work on:
- There is a slight jump when manually dragging the graph (has to do with the equation being applied on line 137 setting the values for `zoom.translateBy`)
    - This causes `mouseup` event not to be triggered, causing the cursor to get stuck in the “grabbed” position
- Graph overflows from bounding box when going from a larger viewport size to smaller viewport size when “Restrict Graph to Viewport” is selected. Need to look into how to apply the `translateExtent` in this case.

**Comment by yshin4** [comment]
Discovered that graph overflow fixed itself (somehow?). Working on the jump.

**Comment by dondi** [comment]
This was looked at by @dondi @yshin4 and @eileenchoe a little more and it was determined that the jump is determined the offset that is being added to the `d3.event` coordinates when the zoom drag is detected. @eileenchoe attempted to eliminate the jump by calculating the local coordinates of the event in relation to the graph bounding box, but this now somehow triggers a strange oscillating behavior where the transform appears to get modified to a different value between mouse events.

This needs further investigation.

**Comment by dondi** [comment]
Something to try for zoom oscillating bug:
- When the mouse goes down, read the transform _then_.
- Save the offset to local variables at that time.
- During the mouse drags, instead of re-reading (and re-parsing) the transform, just use the local variables that were originally read.
This may save the constant reads....just something to try.

**Comment by eileenchoe** [comment]
I fixed the dragging behavior! Here's how I got it to work:
- Bind zoom drag event to the `svg` scope instead of the `boundingBox` scope
- The new binding allowed me to do basically what @dondi proposed above, but without the need to parse the transform at all. The value of the mouse location is saved when a `dragstart` event is detected locally. Then, the difference from the saved location to the new location (which are effectively the x & y transform values) is calculated on `dragged` events.
- Accounted for the effects of scale on the transform offsets
- Fixed cursor grabbing bugs

**Comment by eileenchoe** [comment]
<img width="1077" alt="screen shot 2017-11-29 at 11 29 46 pm" src="https://user-images.githubusercontent.com/16827883/33418969-3d63e3f0-d55d-11e7-8784-778512af167e.png">
This seems to be the last bug to fix for this issue.

**Comment by yshin4** [comment]
I tried doing restrict graph to viewport, zoom less than 100 and larger to smaller but I could not trigger it, trying to figure out when it happens for sure.

**Comment by dondi** [comment]
@eileenchoe and @yshin4 will need to sit together to find a mutually reproducible scenario that manifests the bug. Then we'll have to take it from there.

**Comment by dondi** [comment]
@eileenchoe says that she actually couldn't replicate this bug either. So perhaps @eileenchoe and @yshin4 can have a timeboxed session (plus some code review) seeking to replicate it, and if not, we can move on.

**Comment by yshin4** [comment]
@eileenchoe I went through the file and checked for any undeleted comments and variable names and it seems fine. Seems ready to be merged except the going out of bounding box bug.

**Comment by eileenchoe** [comment]
Couldn't replicate bounding box error either. Opened PR #552

**Comment by eileenchoe** [comment]
@dondi can you review the pull request #552 please?

**Comment by dondi** [comment]
OK @eileenchoe a review is in—looks pretty good but for a few refinements.

**Comment by eileenchoe** [comment]
Merged to beta v2.2.9

**Comment by kdahlquist** [comment]
Closing since the PR is in.
